chemprop/cli/fingerprint.py

```diff
@@ -16,10 +22,190 @@ class FingerprintSubcommand(Subcommand):
+        model_path, multicomponent
+    )  # TODO: connect input_scalers and output_scaler to the model
+    model.eval()
+
+    bounded = any(
+        isinstance(model.criterion, LossFunctionRegistry[loss_function])
+        for loss_function in LossFunctionRegistry.keys()
+        if "bounded" in loss_function
+    )
+
+    format_kwargs = dict(
+        no_header_row=args.no_header_row,
+        smiles_cols=args.smiles_columns,
+        rxn_cols=args.reaction_columns,
+        target_cols=None,
+        ignore_cols=None,
+        weight_col=None,
+        bounded=bounded,
+    )
+    featurization_kwargs = dict(
+        features_generators=args.features_generators, keep_h=args.keep_h, add_h=args.add_h
+    )
+
+    test_data = build_data_from_files(
+        args.test_path,
+        **format_kwargs,
+        p_descriptors=args.descriptors_path,
+        p_atom_feats=args.atom_features_path,
+        p_bond_feats=args.bond_features_path,
+        p_atom_descs=args.atom_descriptors_path,
+        **featurization_kwargs,
+    )
+    logger.info(f"test size: {len(test_data[0])}")
+    test_dsets = [
+        make_dataset(d, args.rxn_mode, args.multi_hot_atom_featurizer_mode) for d in test_data
+    ]
+
+    if multicomponent:
+        test_dset = data.MulticomponentDataset(test_dsets)
+    else:
+        test_dset = test_dsets[0]
+
+    X_d_scaler = input_scalers.get("X_d", None) if input_scalers else None
+    V_f_scaler = input_scalers.get("V_f", None) if input_scalers else None
+    E_f_scaler = input_scalers.get("E_f", None) if input_scalers else None
+    V_d_scaler = input_scalers.get("V_d", None) if input_scalers else None
+
+    if X_d_scaler is not None:
+        test_dset.normalize_inputs("X_d", X_d_scaler)
+    if V_f_scaler is not None:
+        test_dset.normalize_inputs("V_f", V_f_scaler)
+    if E_f_scaler is not None:
+        test_dset.normalize_inputs("E_f", E_f_scaler)
+    if V_d_scaler is not None:
+        test_dset.normalize_inputs("V_d", V_d_scaler)
+
+    test_loader = data.MolGraphDataLoader(
+        test_dset, args.batch_size, args.num_workers, shuffle=False
+    )
+
+    logger.info(model)
+
+    n_layers = len(model.predictor.ffn) - 1
+    if args.ffn_block_index > n_layers:
+        raise ValueError(
+            f"The argument of `ffn_block_index` should not be larger then the `n_layers` of {n_layers} in predictor."
+        )
+
+    with torch.no_grad():
+        if multicomponent:
+            encodings = [
+                model.encoding(batch.bmgs, batch.V_ds, batch.X_d, args.ffn_block_index)
+                for batch in test_loader
+            ]
+        else:
+            encodings = [
+                model.encoding(batch.bmg, batch.V_d, batch.X_d, args.ffn_block_index)
+                for batch in test_loader
+            ]
+        H = torch.cat(encodings, 0)
+
+    if output_path.suffix in [".pkl", ".pckl", ".pickle"]:
+        with open(output_path, "wb") as f:
+            pickle.dump(H, f)
```

Comment:
why pickle and not .npz? We should not be saving tensors as they're loss optimized for I/O than numpy arrays

Reply:
Thanks for the input. I just added an alternative to save as an .npz file. Do we want to keep the pickle file?

Reply:
I think pickle is not necessary. Think of the use-cases for saving encodings:
- plotting with `matplotlib`
- training an `sklearn` model
- used as non-differentiable input to another `torch` model

2/3 use-cases rely on numpy arrays and the third would involve a simple `torch.from_numpy()` call. I think it's clear that the best format to save/load is a numpy array, at which point we should use `.npz` serialization format as it's optimized for that data type.

Reply:
also consider the client usage:
- pickle:
    ```python
    with open(FILE, "rb") as f:
        encodings = pickle.load(f) # encodings is of unknown type
    ```
- `.npz`:
    ```python
    encodings = np.load(FILE) # encodings is of type `np.ndarray`
    ```

It's much cleaner to both load _and_ save using `.npz`